librz/analysis/esil/esil_trace.c

```diff
@@ -18,6 +18,172 @@ static void htup_vector_free(HtUPKv *kv) {
 	rz_vector_free(kv->value);
 }
 
+/**
+ * Get Vector from esil trace db
+ * @param db HtPP* esil->trace->ht_db
+ * @param key const char* behavior string (read/write)
+ * @return vec RzPVector of char *, contains the address of read/write
+ */
+RZ_API RzPVector *rz_analysis_esil_trace_db_get(HtPP *db, const char *key) {
+	RzPVector *vec = ht_pp_find(db, key, NULL);
+
+	if (!vec) {
+		return NULL;
+	}
+
+	return vec;
+}
+
+static void htpp_vector_free(HtPPKv *kv) {
+	rz_vector_free(kv->value);
+}
+
+/**
+ * Add a new place of read/write to vector
+ * @param db HtPP* esil->trace->ht_db
+ * @param key const char* behavior string (read/write)
+ * @param val const char* address/reg/idx
+ */
+RZ_API void rz_analysis_esil_trace_db_array_add(HtPP *db, const char *key, const char *val) {
+	RzPVector *array = rz_analysis_esil_trace_db_get(db, key);
+	if (!array) {
+		// no corresponding value, create one in hash table
+		rz_analysis_esil_trace_db_set(db, key, val);
+		return;
+	}
+
+	if (rz_analysis_esil_trace_db_array_contains(db, key, val)) {
```

Comment:
This means that `array` is retrieve, again, from the HT, when we already have it.